crates/accelerate/src/unitary_compose.rs

```diff
@@ -0,0 +1,149 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use ndarray::{Array, Array2, ArrayView, ArrayView2, IxDyn};
+use ndarray_einsum_beta::*;
+use num_complex::{Complex, Complex64};
+
+static LOWERCASE: [u8; 26] = [
+    b'a', b'b', b'c', b'd', b'e', b'f', b'g', b'h', b'i', b'j', b'k', b'l', b'm', b'n', b'o', b'p',
+    b'q', b'r', b's', b't', b'u', b'v', b'w', b'x', b'y', b'z',
+];
+
+static _UPPERCASE: [u8; 26] = [
+    b'A', b'B', b'C', b'D', b'E', b'F', b'G', b'H', b'I', b'J', b'K', b'L', b'M', b'N', b'O', b'P',
+    b'Q', b'R', b'S', b'T', b'U', b'V', b'W', b'X', b'Y', b'Z',
+];
+
+// Compose the operators given by `gate_unitary` and `overall_unitary`, i.e. apply one to the other
+// as specified by the involved qubits given in `qubits` and the `front` parameter
+pub fn compose(
+    gate_unitary: &ArrayView2<Complex<f64>>,
+    overall_unitary: &ArrayView2<Complex<f64>>,
+    qubits: &[u32],
+    front: bool,
+) -> Array2<Complex<f64>> {
+    let gate_qubits = gate_unitary.shape()[0].ilog2() as usize;
+
+    // Full composition of operators
+    if qubits.is_empty() {
+        if front {
+            return gate_unitary.dot(overall_unitary);
+        } else {
+            return overall_unitary.dot(gate_unitary);
+        }
+    }
+    // Compose with other on subsystem
+    let num_indices = gate_qubits;
+    let shift = if front { gate_qubits } else { 0usize };
+    let right_mul = front;
+
+    //Reshape current matrix
+    //Note that we must reverse the subsystem dimension order as
+    //qubit 0 corresponds to the right-most position in the tensor
+    //product, which is the last tensor wire index.
+    let tensor = per_qubit_shaped(gate_unitary);
+    let mat = per_qubit_shaped(overall_unitary);
+    let indices = qubits
+        .iter()
+        .map(|q| num_indices - 1 - *q as usize)
+        .collect::<Vec<usize>>();
+    let num_rows = usize::pow(2, num_indices as u32);
+
+    _einsum_matmul(&tensor, &mat, &indices, shift, right_mul)
+        .as_standard_layout()
+        .into_shape((num_rows, num_rows))
+        .unwrap()
+        .into_dimensionality::<ndarray::Ix2>()
+        .unwrap()
+        .to_owned()
+}
```

Comment:
The other thing I realized is that we have optimized versions of this for 1 and 2 qubit unitaries already that we use for `optimize1gatesdecomosition` and consolidate blocks. It might be faster to just save the `einsum_matmul` usage for when we have a gate that's >= 3 qubits and reuse are existing composition code for the 1 and 2q case. It would also be good because I think during a typical transpile we're really only going to encounter 1 and 2q gates since we normally only run commutation analysis after we've unrolled larger gates.

Reply:
I've tried custom 1+2 qubit matrix muliplications, but they didn't seem to make much of a difference. I'll keep the code in a custom branch but didn't include it for now, since I couldn't notice a performance improvement on the utility scale benchmarks.

Reply:
Edit: The code now has a special path for 1+2 qubit matrices that directly checks whether they commute, instead of  doing the explicit matrix products (`AB - BA`). It's now significantly faster, as we need less allocations 🙂